source/validate_types.cpp

```diff
@@ -335,36 +359,54 @@ SpvMemoryModel ValidationState_t::getMemoryModel() const {
   return memory_model_;
 }
 
-Functions::Functions(ValidationState_t& module)
-    : module_(module), in_function_(false), in_block_(false) {}
-
-bool Functions::in_function_body() const { return in_function_; }
-
-bool Functions::in_block() const { return in_block_; }
+Function::Function(uint32_t id, uint32_t result_type_id,
+                   SpvFunctionControlMask function_control,
+                   uint32_t function_type_id, ValidationState_t& module)
+    : module_(module),
+      id_(id),
+      function_type_id_(function_type_id),
+      result_type_id_(result_type_id),
+      function_control_(function_control),
+      declaration_type_(FunctionDecl::kFunctionDeclUnknown),
+      blocks_(),
+      current_block_(nullptr),
+      cfg_constructs_(),
+      variable_ids_(),
+      parameter_ids_() {}
+
+bool Function::in_block() const { return static_cast<bool>(current_block_); }
+
+bool Function::IsFirstBlock(uint32_t id) const {
+  return *get_first_block() == id;
+}
 
-spv_result_t Functions::RegisterFunction(uint32_t id, uint32_t ret_type_id,
-                                         uint32_t function_control,
-                                         uint32_t function_type_id) {
+spv_result_t ValidationState_t::RegisterFunction(
+    uint32_t id, uint32_t ret_type_id, SpvFunctionControlMask function_control,
+    uint32_t function_type_id) {
+  assert(in_function_ == false &&
+         "Function instructions can not be declared in a function");
```

Comment:
In that case, I find the assert messages confusing.  They read as if there's a problem with the SPIR-V module and not the code's own contract.